loopy/transform/privatize.py

```diff
@@ -246,4 +248,137 @@ def privatize_temporaries_with_inames(
 # }}}
 
 
+# {{{ unprivatize temporaries with iname
+
+class _InameRemover(IdentityMapper):
+    def __init__(self, inames_to_remove, only_var_names):
+        self.only_var_names = only_var_names
+        self.inames_to_remove = inames_to_remove
+        self.var_name_to_remove_indices = {}
+        super().__init__()
+
+    def map_subscript(self, expr, in_subscript=False):
+        name = expr.aggregate.name
+        if not self.only_var_names or name in self.only_var_names:
+            index = expr.index
+            if not isinstance(index, tuple):
+                index = (index,)
+
+            remove_indices = {}
+            new_index = []
+            for i, index_expr in enumerate(index):
+                if isinstance(index_expr, pymbolic.primitives.Variable) and \
+                        index_expr.name in self.inames_to_remove:
+                    remove_indices[i] = index_expr.name
+                else:
+                    new_index.append(index_expr)
+
+            if name in self.var_name_to_remove_indices:
+                old_remove_indices = self.var_name_to_remove_indices[name]
+                if old_remove_indices != remove_indices:
+                    raise LoopyError(f"Cannot remove indices {remove_indices}"
+                         f" for subscript {expr} because there was another"
+                         f" subscript that required removing indices"
+                         f" {old_remove_indices}")
+            else:
+                self.var_name_to_remove_indices[name] = remove_indices
+
+            if new_index:
+                return expr.aggregate.index(tuple(new_index))
+            else:
+                return expr.aggregate
+        else:
+            return IdentityMapper.map_subscript(self, expr, in_subscript=False)
+
+
+@for_each_kernel
+def unprivatize_temporaries_with_inames(
+        kernel, privatizing_inames, only_var_names=None):
+    """This function reversese the affects of privatize_temporaries_with_inames
+    and removes the private entries in the temporaries each loop iteration of the
+    *privatizing_inames* accesses (possibly restricted to *only_var_names*).
+
+    Example::
+
+        for imatrix, i
+            acc[imatrix] = 0
+            for k
+                acc[imatrix] = acc[imatrix] + a[imatrix, i, k] * vec[k]
+            end
+        end
+
+    might become::
+
+        for imatrix, i
+            acc[0] = 0
```

Comment:
Why is there a spurious `[0]`?

Reply:
Fixed in 009a42435eea3a08c048de73b922481a3c4b1b96